lib/report.js

```diff
@@ -158,7 +160,117 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.failedTests = [];
+    this.testNumeration = 0;
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(
+      test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`
+    );
+  }
+
+  parseTestResults(test, offset = 0) {
+    if (offset === 0) this.testNumeration++;
+    test.results.forEach((res, index) => {
+      const testId = index + 1;
+      if (res.type === 'subtest') {
+        const t = res.test;
+        console.log(`${pad(offset + 2)}# Subtest: ${t.caption}`);
+        this.parseTestResults(t, offset + 2);
+        const prefix = t.success ? 'ok' : 'not ok';
+        console.log(`${pad(offset + 2)}${prefix} ${testId} - ${t.caption}`);
+      } else if (res.success) {
+        this.checkSuccess(res, testId, offset + 2);
+      } else {
+        this.listFailure(res, testId, offset + 2);
+      }
+    });
+    console.log(`${pad(offset + 2)}1..${test.results.length}`);
```

Comment:
```suggestion
```
According to [TAP specification](https://testanything.org/tap-version-13-specification.html):
> The plan cannot appear in the middle of the output, nor can it appear more than once.

And now the plan is printed for every test and subtest which shouldn't be the case.

Also, I'd prefer to print the plan at the beginning.

Reply:
https://www.node-tap.org/subtests/ have a look